ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java

```diff
@@ -265,13 +267,15 @@ private static String idWatermark(CompactionInfo ci) {
   }
 
   /**
-   * @return true if any files were removed
+   * @return true if the cleaner has removed all files rendered obsolete by compaction
    */
   private boolean removeFiles(String location, ValidWriteIdList writeIdList, CompactionInfo ci)
       throws IOException, NoSuchObjectException, MetaException {
     Path locPath = new Path(location);
+    FileSystem fs = locPath.getFileSystem(conf);
+    Map<Path, AcidUtils.HdfsDirSnapshot> dirSnapshots = AcidUtils.getHdfsDirSnapshots(fs, locPath);
     AcidUtils.Directory dir = AcidUtils.getAcidState(locPath.getFileSystem(conf), locPath, conf, writeIdList, Ref.from(
```

Comment:
could be replaced with above fs variable

Reply:
what would happen if there is a long running read-only txn + multiple compaction attempts? are we going to have multiple records in a queue for the same tab;e/partition pending cleanup?

Reply:
No, not with HIVE-24291.
Without HIVE-24291 (which might not be usable if for example if HMS schema changes are out the question) and without this change, we still could have a pileup of the same table/partition in "ready for cleaning" in the queue.
Without this change (HIVE-24444) some of them might not be deleted.
The goal of this change is that, when the table does get cleaned, all of the records will be deleted.